// @cognima/banners/src/album-card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    wrapText,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class AlbumCard
 * @classdesc Generates a visually appealing card displaying album information.
 * @example const card = new AlbumCard()
 * .setAlbumName("Album Title")
 * .setArtistName("Artist Name")
 * .setReleaseYear("2024") // Optional
 * .setTrackCount(12) // Optional
 * .setImage("album_cover_url.png")
 * .build();
 */
module.exports = class AlbumCard {
  constructor(options) {
    // Core Data
    this.albumName = "Album Name";
    this.artistName = "Artist Name";
    this.releaseYear = null;
    this.trackCount = null;
    this.image = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.backgroundColor = "#1D1D1D";
    this.imagePlaceholderColor = "#282828";
    this.albumNameColor = "#FFFFFF";
    this.artistNameColor = "#B3B3B3";
    this.detailsColor = "#B3B3B3";
    this.cardBorderColor = null;
  }

  // --- Setters for Core Data ---
  setAlbumName(name) { if (!name || typeof name !== "string") throw new Error("Album name must be a non-empty string."); this.albumName = name; return this; }
  setArtistName(name) { if (!name || typeof name !== "string") throw new Error("Artist name must be a non-empty string."); this.artistName = name; return this; }
  setReleaseYear(year) { if (year && typeof year !== "string" && typeof year !== "number") throw new Error("Release year must be a string or number if provided."); this.releaseYear = year.toString(); return this; }
  setTrackCount(count) { if (count && typeof count !== "number") throw new Error("Track count must be a number if provided."); if (count <= 0) throw new Error("Track count must be positive if provided."); this.trackCount = count; return this; }
  setImage(image) { if (!image) throw new Error("Image source cannot be empty."); this.image = image; return this; }

  // --- Setters for Visual Customization ---
  setBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid background color. Use hex format."); this.backgroundColor = color; return this; }
  setImagePlaceholderColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid image placeholder color. Use hex format."); this.imagePlaceholderColor = color; return this; }
  setAlbumNameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid album name color. Use hex format."); this.albumNameColor = color; return this; }
  setArtistNameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid artist name color. Use hex format."); this.artistNameColor = color; return this; }
  setDetailsColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid details color. Use hex format."); this.detailsColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }

  // --- Build Method ---
  async build() {
    // --- Validation ---
    if (!this.image) throw new Error("Image must be set using setImage().");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup ---
    const cardWidth = 300;
    const cardHeight = 420;
    const padding = 20;
    const imageSize = cardWidth - 2 * padding;
    const borderRadius = 15;
    const textPaddingTop = 15;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Background ---
    ctx.fillStyle = this.backgroundColor;
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, true, false);

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor;
      ctx.lineWidth = 4;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Image ---
    const imageX = padding;
    const imageY = padding;
    const imageRadius = 10;
    try {
      const img = await loadImageWithAxios(this.image);
      ctx.save();
      roundRect(ctx, imageX, imageY, imageSize, imageSize, imageRadius, false, false);
      ctx.clip();
      ctx.drawImage(img, imageX, imageY, imageSize, imageSize);
      ctx.restore();
    } catch (e) {
      console.error("Failed to draw album cover image:", e.message);
      ctx.fillStyle = this.imagePlaceholderColor;
      roundRect(ctx, imageX, imageY, imageSize, imageSize, imageRadius, true, false);
    }

    // --- Draw Text ---
    const textStartX = padding;
    const textStartY = imageY + imageSize + textPaddingTop;
    const textWidth = cardWidth - 2 * padding;

    ctx.textAlign = "start";
    ctx.textBaseline = "top";

    // Album Name (Bold, larger)
    ctx.fillStyle = this.albumNameColor;
    const albumFont = `20px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = albumFont;
    const albumText = this.albumName.length > 25 ? this.albumName.slice(0, 22) + "..." : this.albumName;
    ctx.fillText(albumText, textStartX, textStartY);

    // Artist Name (Medium, smaller)
    ctx.fillStyle = this.artistNameColor;
    const artistFont = `16px ${registeredFontName}-Medium`; // Use registered name
    ctx.font = artistFont;
    const artistText = this.artistName.length > 30 ? this.artistName.slice(0, 27) + "..." : this.artistName;
    const albumNameHeight = 26;
    const artistY = textStartY + albumNameHeight;
    ctx.fillText(artistText, textStartX, artistY);

    // Details (Year and Track Count - Regular, smallest)
    let detailsText = "";
    if (this.releaseYear) {
        detailsText += `Album • ${this.releaseYear}`;
    }
    if (this.trackCount) {
        detailsText += (detailsText ? " • " : "") + `${this.trackCount} track${this.trackCount > 1 ? "s" : ""}`;
    }

    if (detailsText) {
        ctx.fillStyle = this.detailsColor;
        const detailsFont = `14px ${registeredFontName}-Regular`; // Use registered name
        ctx.font = detailsFont;
        const artistNameHeight = 20;
        const detailsY = artistY + artistNameHeight;
        ctx.fillText(detailsText, textStartX, detailsY);
    }

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Album Card:", err);
      throw new Error("Could not generate the Album Card image buffer.");
    }
  }
};
