// @cognima/banners/src/now-playing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    formatTime,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class NowPlaying
 * @classdesc Generates a compact and modern "Now Playing" card.
 * @example const card = new NowPlaying()
 * .setAuthor("Artist Name")
 * .setTitle("Song Title")
 * .setAlbum("Album Name") // Optional
 * .setImage("image_url.png")
 * .setTimestamp(60000, 240000) // 1:00 / 4:00
 * .build();
 */
module.exports = class NowPlaying {
  constructor(options) {
    // Core Data
    this.author = null;
    this.title = null;
    this.album = null; // Optional
    this.image = null;
    this.start = null;
    this.end = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.backgroundColor = "#2F3136";
    this.imagePlaceholderColor = "#3C3F44";
    this.authorColor = "#FFFFFF";
    this.titleColor = "#B9BBBE";
    this.albumColor = "#B9BBBE";
    this.timestampColor = "#B9BBBE";
    this.progressBarColor = "#FFFFFF";
    this.progressBarBaseColor = "#484B4E";
    this.cardBorderColor = null;
  }

  // --- Setters for Core Data ---
  setAuthor(name) { if (!name || typeof name !== "string") throw new Error("Author name must be a non-empty string."); this.author = name; return this; }
  setTitle(title) { if (!title || typeof title !== "string") throw new Error("Title must be a non-empty string."); this.title = title; return this; }
  setAlbum(name) { if (name && typeof name !== "string") throw new Error("Album name must be a string if provided."); this.album = name; return this; }
  setImage(image) { if (!image) throw new Error("Image source cannot be empty."); this.image = image; return this; }
  setTimestamp(start, end) { if (typeof start !== "number" || start < 0) throw new Error("Start timestamp must be a non-negative number."); if (typeof end !== "number" || end <= 0) throw new Error("End timestamp must be a positive number."); if (start > end) console.warn("Start timestamp is greater than end timestamp."); this.start = start; this.end = end; return this; }

  // --- Setters for Visual Customization ---
  setBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid background color. Use hex format."); this.backgroundColor = color; return this; }
  setImagePlaceholderColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid image placeholder color. Use hex format."); this.imagePlaceholderColor = color; return this; }
  setAuthorColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid author color. Use hex format."); this.authorColor = color; return this; }
  setTitleColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid title color. Use hex format."); this.titleColor = color; return this; }
  setAlbumColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid album color. Use hex format."); this.albumColor = color; return this; }
  setTimestampColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid timestamp color. Use hex format."); this.timestampColor = color; return this; }
  setProgressBarColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar color. Use hex format."); this.progressBarColor = color; return this; }
  setProgressBarBaseColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar base color. Use hex format."); this.progressBarBaseColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }

  // --- Private Helper Methods ---
  _calculateProgress(barWidth) {
    if (this.end <= 0) return 0;
    const progressRatio = Math.max(0, Math.min(this.start / this.end, 1));
    return progressRatio * barWidth;
  }

  // --- Build Method ---
  async build() {
    // --- Validation ---
    if (!this.author) throw new Error("Author must be set using setAuthor().");
    if (!this.title) throw new Error("Title must be set using setTitle().");
    if (!this.image) throw new Error("Image must be set using setImage().");
    if (this.start === null || this.end === null) throw new Error("Timestamps must be set using setTimestamp().");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup ---
    const cardWidth = 600;
    const cardHeight = 160;
    const imageSize = 120;
    const borderRadius = 10;
    const padding = 20;
    const textPadding = 20;
    const progressBarHeight = 6;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Background ---
    ctx.fillStyle = this.backgroundColor;
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, true, false);

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor;
      ctx.lineWidth = 4;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Image ---
    const imageX = padding;
    const imageY = (cardHeight - imageSize) / 2;
    const imageRadius = 8;
    try {
      const img = await loadImageWithAxios(this.image);
      ctx.save();
      roundRect(ctx, imageX, imageY, imageSize, imageSize, imageRadius, false, false);
      ctx.clip();
      ctx.drawImage(img, imageX, imageY, imageSize, imageSize);
      ctx.restore();
    } catch (e) {
      console.error("Failed to draw artwork image:", e.message);
      ctx.fillStyle = this.imagePlaceholderColor;
      roundRect(ctx, imageX, imageY, imageSize, imageSize, imageRadius, true, false);
    }

    // --- Text and Progress Bar Area ---
    const textStartX = imageX + imageSize + textPadding;
    const textEndX = cardWidth - padding;
    const textWidth = textEndX - textStartX;

    // --- Draw Text ---
    ctx.textAlign = "start";
    ctx.textBaseline = "top";

    // Author (Bold, larger)
    ctx.fillStyle = this.authorColor;
    const authorFont = `22px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = authorFont;
    const authorText = this.author.length > 25 ? this.author.slice(0, 22) + "..." : this.author;
    const authorY = padding;
    ctx.fillText(authorText, textStartX, authorY);

    // Title (Medium, slightly smaller)
    ctx.fillStyle = this.titleColor;
    const titleFont = `18px ${registeredFontName}-Medium`; // Use registered name
    ctx.font = titleFont;
    const titleText = this.title.length > 30 ? this.title.slice(0, 27) + "..." : this.title;
    const authorHeight = 28;
    const titleY = authorY + authorHeight;
    ctx.fillText(titleText, textStartX, titleY);

    // Album (Optional, Regular, smallest)
    let albumHeight = 0;
    if (this.album) {
        ctx.fillStyle = this.albumColor;
        const albumFont = `16px ${registeredFontName}-Regular`; // Use registered name
        ctx.font = albumFont;
        const albumText = this.album.length > 35 ? this.album.slice(0, 32) + "..." : this.album;
        albumHeight = 24;
        const albumY = titleY + albumHeight;
        ctx.fillText(albumText, textStartX, albumY);
    }

    // --- Draw Progress Bar & Timestamps ---
    const barBaseY = cardHeight - padding - progressBarHeight - 15;
    const barWidth = textWidth;
    const barStartX = textStartX;
    const barRadius = progressBarHeight / 2;
    const timestampY = barBaseY + progressBarHeight + 5;

    // Bar background
    ctx.fillStyle = this.progressBarBaseColor;
    roundRect(ctx, barStartX, barBaseY, barWidth, progressBarHeight, barRadius, true, false);

    // Bar progress
    const progressWidth = this._calculateProgress(barWidth);
    if (progressWidth > 0) {
      ctx.fillStyle = this.progressBarColor;
      roundRect(ctx, barStartX, barBaseY, progressWidth, progressBarHeight, barRadius, true, false);
    }

    // Timestamps
    const start_format = formatTime(this.start);
    const end_format = formatTime(this.end);
    ctx.fillStyle = this.timestampColor;
    const timestampFont = `14px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = timestampFont;
    ctx.textBaseline = "top";
    ctx.fillText(start_format, barStartX, timestampY);
    ctx.textAlign = "end";
    ctx.fillText(end_format, barStartX + barWidth, timestampY);

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Now Playing card:", err);
      throw new Error("Could not generate the Now Playing card image buffer.");
    }
  }
};
