// @cognima/banners/src/profile-card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    wrapText,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class ProfileCard
 * @classdesc Generates a user profile card with customizable elements.
 * @example const card = new ProfileCard()
 * .setUsername("CognimaUser")
 * .setAvatar("avatar.png")
 * .setBackground("image", "header.jpg") // Or ("color", "#abcdef")
 * .setBio("AI Assistant | Building cool things with code.")
 * .addBadge({ url: "badge1.png", description: "Early Supporter" })
 * .addBadge({ url: "badge2.png", description: "Verified Bot" })
 * .setCustomField("Joined", "May 2025")
 * .setCustomField("Location", "Cyberspace")
 * .build();
 */
module.exports = class ProfileCard {
  constructor(options) {
    // Core Data
    this.username = "Username";
    this.avatar = null;
    this.bio = null;
    this.badges = [];
    this.customFields = {};

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.background = { type: "color", value: "#2C2F33" };
    this.bodyBackgroundColor = "#23272A";
    this.usernameColor = "#FFFFFF";
    this.bioColor = "#B9BBBE";
    this.fieldTitleColor = "#FFFFFF";
    this.fieldValueColor = "#B9BBBE";
    this.badgeBackgroundColor = "#2F3136";
    this.avatarBorderColor = null;
    this.cardBorderColor = null;
    this.overlayOpacity = 0.4;
    this.status = { type: "offline", color: "#747F8D" };
  }

  // --- Setters for Core Data ---
  setUsername(name) { if (!name || typeof name !== "string") throw new Error("Username must be a non-empty string."); this.username = name; return this; }
  setAvatar(image) { if (!image) throw new Error("Avatar image source cannot be empty."); this.avatar = image; return this; }
  setBio(text) { if (text && typeof text !== "string") throw new Error("Bio must be a string if provided."); this.bio = text; return this; }
  addBadge(badge) { if (!badge || typeof badge !== "object" || !badge.url) throw new Error("Badge must be an object with at least a \"url\" property."); this.badges.push({ url: badge.url, description: badge.description || "Badge" }); return this; }
  setCustomField(title, value) { if (!title || typeof title !== "string") throw new Error("Custom field title must be a non-empty string."); if (!value || typeof value !== "string") throw new Error("Custom field value must be a non-empty string."); this.customFields[title] = value; return this; }

  // --- Setters for Visual Customization ---
  setBackground(type, value) {
    const types = ["color", "image"];
    if (!type || !types.includes(type.toLowerCase())) throw new Error("Background type must be \"color\" or \"image\".");
    if (!value) throw new Error("Background value cannot be empty.");
    if (type.toLowerCase() === "color" && !isValidHexColor(value)) throw new Error("Invalid background color. Use hex format.");
    this.background = { type: type.toLowerCase(), value };
    return this;
  }

  setBodyBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid body background color. Use hex format."); this.bodyBackgroundColor = color; return this; }
  setUsernameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid username color. Use hex format."); this.usernameColor = color; return this; }
  setBioColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid bio color. Use hex format."); this.bioColor = color; return this; }
  setFieldTitleColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid field title color. Use hex format."); this.fieldTitleColor = color; return this; }
  setFieldValueColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid field value color. Use hex format."); this.fieldValueColor = color; return this; }
  setBadgeBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid badge background color. Use hex format."); this.badgeBackgroundColor = color; return this; }
  setAvatarBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid avatar border color. Use hex format."); this.avatarBorderColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }
  setOverlayOpacity(opacity) { if (typeof opacity !== "number" || opacity < 0 || opacity > 1) throw new Error("Overlay opacity must be between 0 and 1."); this.overlayOpacity = opacity; return this; }
  setStatus(type) {
    const validTypes = { online: "#43B581", idle: "#FAA61A", dnd: "#F04747", streaming: "#593695", offline: "#747F8D" };
    if (!type || !validTypes[type.toLowerCase()]) throw new Error(`Invalid status type. Use one of: ${Object.keys(validTypes).join(", ")}`);
    this.status = { type: type.toLowerCase(), color: validTypes[type.toLowerCase()] };
    return this;
  }

  // --- Build Method ---
  async build() {
    if (!this.avatar) throw new Error("Avatar image must be set using setAvatar().");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup ---
    const cardWidth = 600;
    const headerHeight = 180;
    const avatarSize = 128;
    const avatarOverlap = 40;
    const badgeAreaHeight = this.badges.length > 0 ? 80 : 0;
    const bodyPadding = 25;
    const contentStartY = headerHeight - avatarOverlap + avatarSize / 2 + bodyPadding;
    let cardHeight = contentStartY + 100; // Initial height

    // Estimate height needed for bio and fields
    let estimatedContentHeight = 0;
    if (this.bio) estimatedContentHeight += 60;
    estimatedContentHeight += Object.keys(this.customFields).length * 45;
    cardHeight = contentStartY + Math.max(100, estimatedContentHeight) + badgeAreaHeight + bodyPadding;

    const borderRadius = 15;
    const statusIndicatorSize = 32;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Main Body Background ---
    ctx.fillStyle = this.bodyBackgroundColor;
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, true, false);

    // --- Draw Header Background & Overlay ---
    ctx.save();
    // Define the clipping path for the header (top rounded corners only)
    ctx.beginPath();
    ctx.moveTo(0, headerHeight); // Start bottom-left
    ctx.lineTo(0, borderRadius); // Left edge up to radius
    ctx.quadraticCurveTo(0, 0, borderRadius, 0); // Top-left corner
    ctx.lineTo(cardWidth - borderRadius, 0); // Top edge
    ctx.quadraticCurveTo(cardWidth, 0, cardWidth, borderRadius); // Top-right corner
    ctx.lineTo(cardWidth, headerHeight); // Right edge down
    ctx.closePath(); // Close path back to bottom-right (implicitly)
    ctx.clip();

    ctx.globalAlpha = 1;
    if (this.background.type === "color") {
      ctx.fillStyle = this.background.value;
      ctx.fillRect(0, 0, cardWidth, headerHeight);
    } else if (this.background.type === "image") {
      try {
        const img = await loadImageWithAxios(this.background.value);
        const aspect = img.width / img.height;
        let drawWidth = cardWidth; let drawHeight = cardWidth / aspect;
        if (drawHeight < headerHeight) { drawHeight = headerHeight; drawWidth = headerHeight * aspect; }
        const offsetX = (cardWidth - drawWidth) / 2; const offsetY = 0;
        ctx.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
      } catch (e) {
        console.error("Failed to draw header background image:", e.message);
        ctx.fillStyle = "#2C2F33"; ctx.fillRect(0, 0, cardWidth, headerHeight);
      }
    }

    if (this.overlayOpacity > 0 && this.background.type === "image") {
      ctx.globalAlpha = this.overlayOpacity; ctx.fillStyle = "#000000"; ctx.fillRect(0, 0, cardWidth, headerHeight);
    }
    ctx.restore(); ctx.globalAlpha = 1;

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor; ctx.lineWidth = 4;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Avatar ---
    const avatarX = bodyPadding;
    const avatarY = headerHeight - avatarOverlap - avatarSize / 2;

    ctx.save();
    ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2, 0, Math.PI * 2); ctx.closePath(); ctx.clip();
    try {
      const avatarImg = await loadImageWithAxios(this.avatar);
      ctx.drawImage(avatarImg, avatarX, avatarY, avatarSize, avatarSize);
    } catch (e) { /* ... fallback avatar ... */
        console.error("Failed to draw avatar image:", e.message);
        ctx.fillStyle = "#555"; ctx.fillRect(avatarX, avatarY, avatarSize, avatarSize);
        ctx.fillStyle = "#FFF";
        const fallbackFont = `30px ${registeredFontName}-Bold`; // Use registered name
        ctx.font = fallbackFont;
        ctx.textAlign = "center"; ctx.textBaseline = "middle";
        ctx.fillText("?", avatarX + avatarSize / 2, avatarY + avatarSize / 2);
    }
    ctx.restore();

    // --- Draw Avatar Border (Optional) ---
    if (this.avatarBorderColor) {
        ctx.strokeStyle = this.avatarBorderColor; ctx.lineWidth = 5;
        ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2 + ctx.lineWidth / 2, 0, Math.PI * 2); ctx.stroke(); ctx.closePath();
    }

    // --- Draw Status Indicator ---
    ctx.fillStyle = this.status.color;
    ctx.beginPath();
    const statusX = avatarX + avatarSize - statusIndicatorSize * 0.7;
    const statusY = avatarY + avatarSize - statusIndicatorSize * 0.7;
    ctx.arc(statusX, statusY, statusIndicatorSize / 2, 0, Math.PI * 2); ctx.fill(); ctx.closePath();
    ctx.strokeStyle = this.bodyBackgroundColor; ctx.lineWidth = 3; ctx.stroke();

    // --- Draw Username ---
    ctx.fillStyle = this.usernameColor;
    const usernameFont = `28px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = usernameFont;
    ctx.textAlign = "start"; ctx.textBaseline = "top";
    const usernameX = avatarX + avatarSize + bodyPadding;
    const usernameY = headerHeight + 5;
    const usernameText = this.username.length > 20 ? this.username.slice(0, 17) + "..." : this.username;
    ctx.fillText(usernameText, usernameX, usernameY);

    // --- Draw Content Below Avatar ---
    let currentY = contentStartY;
    const contentX = bodyPadding;
    const contentWidth = cardWidth - 2 * bodyPadding;

    // Bio
    if (this.bio) {
        ctx.fillStyle = this.bioColor;
        const bioFont = `18px ${registeredFontName}-Regular`; // Use registered name
        ctx.font = bioFont;
        ctx.textAlign = "start"; ctx.textBaseline = "top";
        currentY = wrapText(ctx, this.bio, contentX, currentY, contentWidth, 24, registeredFontName) + 15;
    }

    // Custom Fields
    if (Object.keys(this.customFields).length > 0) {
        currentY += 10;
        ctx.textBaseline = "top";
        for (const title in this.customFields) {
            ctx.fillStyle = this.fieldTitleColor;
            const fieldTitleFont = `16px ${registeredFontName}-Bold`; // Use registered name
            ctx.font = fieldTitleFont;
            ctx.fillText(title.toUpperCase(), contentX, currentY);

            ctx.fillStyle = this.fieldValueColor;
            const fieldValueFont = `16px ${registeredFontName}-Regular`; // Use registered name
            ctx.font = fieldValueFont;
            const valueText = this.customFields[title];
            ctx.fillText(valueText, contentX, currentY + 22);
            currentY += 45;
        }
    }

    // --- Draw Badges Area ---
    if (this.badges.length > 0) {
        const badgeAreaY = cardHeight - badgeAreaHeight;
        ctx.fillStyle = this.badgeBackgroundColor;
        ctx.fillRect(0, badgeAreaY, cardWidth, badgeAreaHeight);

        const badgeSize = 40;
        const badgePadding = 15;
        let currentBadgeX = bodyPadding;
        const badgeY = badgeAreaY + (badgeAreaHeight - badgeSize) / 2;

        for (const badge of this.badges.slice(0, 10)) {
            try {
                const badgeImg = await loadImageWithAxios(badge.url);
                ctx.drawImage(badgeImg, currentBadgeX, badgeY, badgeSize, badgeSize);
                currentBadgeX += badgeSize + badgePadding;
            } catch (e) {
                console.warn(`Failed to load badge image: ${badge.url}`, e.message);
                ctx.fillStyle = "#555"; ctx.fillRect(currentBadgeX, badgeY, badgeSize, badgeSize);
                currentBadgeX += badgeSize + badgePadding;
            }
        }
    }

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Profile Card:", err);
      throw new Error("Could not generate the Profile Card image buffer.");
    }
  }
};
