// @cognima/banners/src/rank-v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    formatNumber,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class RankV2
 * @classdesc Generates a compact, alternative rank card design.
 * @example const card = new RankV2()
 * .setUsername("Username")
 * .setAvatar("avatar.png")
 * .setCurrentXp(750)
 * .setRequiredXp(1000)
 * .setLevel(15)
 * .setRank(3)
 * .build();
 */
module.exports = class RankV2 {
  constructor(options) {
    // Core Data
    this.username = "Username";
    this.avatar = null;
    this.currentXp = 0;
    this.requiredXp = 100;
    this.level = 1;
    this.rank = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.background = { type: "color", value: "#2A2E35" };
    this.progressBarColor = "#FFFFFF";
    this.progressBarBaseColor = "#484B4E";
    this.usernameColor = "#FFFFFF";
    this.levelColor = "#FFFFFF";
    this.rankColor = "#B9BBBE";
    this.xpColor = "#B9BBBE";
    this.avatarBorderColor = null;
    this.cardBorderColor = null;
    this.overlayOpacity = 0;
    this.status = { type: "offline", color: "#747F8D" };
  }

  // --- Setters for Core Data ---
  setUsername(name) { if (!name || typeof name !== "string") throw new Error("Username must be a non-empty string."); this.username = name; return this; }
  setAvatar(image) { if (!image) throw new Error("Avatar image source cannot be empty."); this.avatar = image; return this; }
  setCurrentXp(xp) { if (typeof xp !== "number" || xp < 0) throw new Error("Current XP must be a non-negative number."); this.currentXp = xp; return this; }
  setRequiredXp(xp) { if (typeof xp !== "number" || xp <= 0) throw new Error("Required XP must be a positive number."); this.requiredXp = xp; return this; }
  setLevel(level) { if (typeof level !== "number" || level < 0) throw new Error("Level must be a non-negative number."); this.level = level; return this; }
  setRank(rank) { if (rank !== null && (typeof rank !== "number" || rank < 0)) throw new Error("Rank must be a non-negative number or null."); this.rank = rank; return this; }

  // --- Setters for Visual Customization ---
  setBackground(type, value) {
    const types = ["color", "image"];
    if (!type || !types.includes(type.toLowerCase())) throw new Error("Background type must be \'color\' or \'image\'.");
    if (!value) throw new Error("Background value cannot be empty.");
    if (type.toLowerCase() === "color" && !isValidHexColor(value)) throw new Error("Invalid background color. Use hex format.");
    this.background = { type: type.toLowerCase(), value };
    return this;
  }

  setProgressBarColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar color. Use hex format."); this.progressBarColor = color; return this; }
  setProgressBarBaseColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar base color. Use hex format."); this.progressBarBaseColor = color; return this; }
  setUsernameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid username color. Use hex format."); this.usernameColor = color; return this; }
  setLevelColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid level color. Use hex format."); this.levelColor = color; return this; }
  setRankColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid rank color. Use hex format."); this.rankColor = color; return this; }
  setXpColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid XP color. Use hex format."); this.xpColor = color; return this; }
  setAvatarBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid avatar border color. Use hex format."); this.avatarBorderColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }
  setOverlayOpacity(opacity) { if (typeof opacity !== "number" || opacity < 0 || opacity > 1) throw new Error("Overlay opacity must be between 0 and 1."); this.overlayOpacity = opacity; return this; }
  setStatus(type) {
    const validTypes = { online: "#43B581", idle: "#FAA61A", dnd: "#F04747", streaming: "#593695", offline: "#747F8D" };
    if (!type || !validTypes[type.toLowerCase()]) throw new Error(`Invalid status type. Use one of: ${Object.keys(validTypes).join(", ")}`);
    this.status = { type: type.toLowerCase(), color: validTypes[type.toLowerCase()] };
    return this;
  }

  // --- Private Helper Methods ---
  _calculateProgress(barWidth) {
    const progress = (this.currentXp / this.requiredXp) * barWidth;
    return Math.max(0, Math.min(progress, barWidth));
  }

  // --- Build Method (V2 Layout) ---
  async build() {
    if (!this.avatar) throw new Error("Avatar image must be set using setAvatar().");
    if (this.currentXp > this.requiredXp) console.warn("Current XP exceeds required XP.");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup (Compact Design) ---
    const cardWidth = 700;
    const cardHeight = 180;
    const avatarSize = 120;
    const padding = 25;
    const avatarX = padding;
    const avatarY = (cardHeight - avatarSize) / 2;
    const borderRadius = 15;
    const statusIndicatorSize = 30;
    const progressBarHeight = 10;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Background & Overlay ---
    ctx.save();
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, false, false);
    ctx.clip();
    ctx.globalAlpha = 1;
    if (this.background.type === "color") {
      ctx.fillStyle = this.background.value;
      ctx.fillRect(0, 0, cardWidth, cardHeight);
    } else { /* ... image background logic ... */
        try {
            const img = await loadImageWithAxios(this.background.value);
            const aspect = img.width / img.height;
            let drawWidth = cardWidth; let drawHeight = cardWidth / aspect;
            if (drawHeight < cardHeight) { drawHeight = cardHeight; drawWidth = cardHeight * aspect; }
            const offsetX = (cardWidth - drawWidth) / 2; const offsetY = (cardHeight - drawHeight) / 2;
            ctx.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
        } catch (e) {
            console.error("Failed to draw background image:", e.message);
            ctx.fillStyle = "#2A2E35"; ctx.fillRect(0, 0, cardWidth, cardHeight);
        }
    }
    if (this.overlayOpacity > 0 && this.background.type === "image") {
      ctx.globalAlpha = this.overlayOpacity; ctx.fillStyle = "#000000"; ctx.fillRect(0, 0, cardWidth, cardHeight);
    }
    ctx.restore(); ctx.globalAlpha = 1;

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
        ctx.strokeStyle = this.cardBorderColor; ctx.lineWidth = 4;
        roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Avatar ---
    ctx.save();
    ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2, 0, Math.PI * 2); ctx.closePath(); ctx.clip();
    try {
      const avatarImg = await loadImageWithAxios(this.avatar);
      ctx.drawImage(avatarImg, avatarX, avatarY, avatarSize, avatarSize);
    } catch (e) { /* ... fallback avatar ... */
        console.error("Failed to draw avatar image:", e.message);
        ctx.fillStyle = "#555"; ctx.fillRect(avatarX, avatarY, avatarSize, avatarSize);
        ctx.fillStyle = "#FFF";
        const fallbackFont = `30px ${registeredFontName}-Bold`; // Use registered name
        ctx.font = fallbackFont;
        ctx.textAlign = "center"; ctx.textBaseline = "middle";
        ctx.fillText("?", avatarX + avatarSize / 2, avatarY + avatarSize / 2);
    }
    ctx.restore();

    // --- Draw Avatar Border (Optional) ---
    if (this.avatarBorderColor) {
        ctx.strokeStyle = this.avatarBorderColor; ctx.lineWidth = 4;
        ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2 + ctx.lineWidth / 2, 0, Math.PI * 2); ctx.stroke(); ctx.closePath();
    }

    // --- Draw Status Indicator ---
    ctx.fillStyle = this.status.color;
    ctx.beginPath();
    const statusX = avatarX + avatarSize - statusIndicatorSize * 0.8;
    const statusY = avatarY + avatarSize - statusIndicatorSize * 0.8;
    ctx.arc(statusX, statusY, statusIndicatorSize / 2, 0, Math.PI * 2); ctx.fill(); ctx.closePath();
    ctx.strokeStyle = this.background.type === "color" ? this.background.value : "#2A2E35"; // Use background color for border
    ctx.lineWidth = 3; ctx.stroke();

    // --- Draw Text and Progress Bar Area (Compact Layout) ---
    const textStartX = avatarX + avatarSize + padding;
    const textEndX = cardWidth - padding;
    const textAreaWidth = textEndX - textStartX;

    // Username and Rank (Top Line)
    ctx.fillStyle = this.usernameColor;
    const usernameFont = `26px ${registeredFontName}-Medium`; // Use registered name
    ctx.font = usernameFont;
    ctx.textAlign = "start";
    ctx.textBaseline = "top";
    const usernameText = this.username.length > 18 ? this.username.slice(0, 15) + "..." : this.username;
    const topTextY = avatarY + 5;
    ctx.fillText(usernameText, textStartX, topTextY);

    if (this.rank !== null) {
        ctx.fillStyle = this.rankColor;
        const rankFont = `22px ${registeredFontName}-Medium`; // Use registered name
        ctx.font = rankFont;
        ctx.textAlign = "end";
        const rankText = `Rank #${formatNumber(this.rank)}`;
        ctx.fillText(rankText, textEndX, topTextY + 4);
    }

    // Level (Below Username)
    ctx.fillStyle = this.levelColor;
    const levelFont = `20px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = levelFont;
    ctx.textAlign = "start";
    const levelText = `Level ${formatNumber(this.level)}`;
    const levelY = topTextY + 35;
    ctx.fillText(levelText, textStartX, levelY);

    // XP Text (Right Aligned, below Rank)
    ctx.fillStyle = this.xpColor;
    const xpFont = `18px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = xpFont;
    ctx.textAlign = "end";
    const xpText = `${formatNumber(this.currentXp)} / ${formatNumber(this.requiredXp)} XP`;
    ctx.fillText(xpText, textEndX, levelY + 2);

    // Progress Bar (Bottom)
    const progressBarY = avatarY + avatarSize - progressBarHeight;
    const barWidth = textAreaWidth;
    const barRadius = progressBarHeight / 2;

    ctx.fillStyle = this.progressBarBaseColor;
    roundRect(ctx, textStartX, progressBarY, barWidth, progressBarHeight, barRadius, true, false);

    const progressWidth = this._calculateProgress(barWidth);
    if (progressWidth > 0) {
      ctx.fillStyle = this.progressBarColor;
      roundRect(ctx, textStartX, progressBarY, progressWidth, progressBarHeight, barRadius, true, false);
    }

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Rank V2 card:", err);
      throw new Error("Could not generate the Rank V2 card image buffer.");
    }
  }
};
