// @cognima/banners/src/rank.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    formatNumber,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class Rank
 * @classdesc Generates a modern and visually appealing rank card for users.
 * @example const card = new Rank()
 * .setUsername("Username")
 * .setAvatar("avatar.png")
 * .setCurrentXp(750)
 * .setRequiredXp(1000)
 * .setLevel(15)
 * .setRank(3)
 * .build();
 */
module.exports = class Rank {
  constructor(options) {
    // Core Data
    this.username = "Username";
    this.avatar = null;
    this.currentXp = 0;
    this.requiredXp = 100;
    this.level = 1;
    this.rank = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.background = { type: "color", value: "#23272A" };
    this.progressBarColor = "#FFFFFF";
    this.progressBarBaseColor = "#484B4E";
    this.usernameColor = "#FFFFFF";
    this.levelColor = "#FFFFFF";
    this.rankColor = "#B9BBBE";
    this.xpColor = "#B9BBBE";
    this.avatarBorderColor = null;
    this.cardBorderColor = null;
    this.overlayOpacity = 0;
    this.status = { type: "offline", color: "#747F8D" };
  }

  // --- Setters for Core Data ---
  setUsername(name) { if (!name || typeof name !== "string") throw new Error("Username must be a non-empty string."); this.username = name; return this; }
  setAvatar(image) { if (!image) throw new Error("Avatar image source cannot be empty."); this.avatar = image; return this; }
  setCurrentXp(xp) { if (typeof xp !== "number" || xp < 0) throw new Error("Current XP must be a non-negative number."); this.currentXp = xp; return this; }
  setRequiredXp(xp) { if (typeof xp !== "number" || xp <= 0) throw new Error("Required XP must be a positive number."); this.requiredXp = xp; return this; }
  setLevel(level) { if (typeof level !== "number" || level < 0) throw new Error("Level must be a non-negative number."); this.level = level; return this; }
  setRank(rank) { if (rank !== null && (typeof rank !== "number" || rank < 0)) throw new Error("Rank must be a non-negative number or null."); this.rank = rank; return this; }

  // --- Setters for Visual Customization ---
  setBackground(type, value) {
    const types = ["color", "image"];
    if (!type || !types.includes(type.toLowerCase())) throw new Error("Background type must be \'color\' or \'image\'.");
    if (!value) throw new Error("Background value cannot be empty.");
    if (type.toLowerCase() === "color" && !isValidHexColor(value)) throw new Error("Invalid background color. Use hex format.");
    this.background = { type: type.toLowerCase(), value };
    return this;
  }

  setProgressBarColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar color. Use hex format."); this.progressBarColor = color; return this; }
  setProgressBarBaseColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar base color. Use hex format."); this.progressBarBaseColor = color; return this; }
  setUsernameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid username color. Use hex format."); this.usernameColor = color; return this; }
  setLevelColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid level color. Use hex format."); this.levelColor = color; return this; }
  setRankColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid rank color. Use hex format."); this.rankColor = color; return this; }
  setXpColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid XP color. Use hex format."); this.xpColor = color; return this; }
  setAvatarBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid avatar border color. Use hex format."); this.avatarBorderColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }
  setOverlayOpacity(opacity) { if (typeof opacity !== "number" || opacity < 0 || opacity > 1) throw new Error("Overlay opacity must be between 0 and 1."); this.overlayOpacity = opacity; return this; }
  setStatus(type) {
    const validTypes = { online: "#43B581", idle: "#FAA61A", dnd: "#F04747", streaming: "#593695", offline: "#747F8D" };
    if (!type || !validTypes[type.toLowerCase()]) throw new Error(`Invalid status type. Use one of: ${Object.keys(validTypes).join(", ")}`);
    this.status = { type: type.toLowerCase(), color: validTypes[type.toLowerCase()] };
    return this;
  }

  // --- Private Helper Methods ---
  _calculateProgress(barWidth) {
    const progress = (this.currentXp / this.requiredXp) * barWidth;
    return Math.max(0, Math.min(progress, barWidth));
  }

  // --- Build Method ---
  async build() {
    console.log("--- Rank Card Build Start ---");
    if (!this.avatar) throw new Error("Avatar image must be set using setAvatar().");
    if (this.currentXp > this.requiredXp) console.warn("Current XP exceeds required XP.");

    // --- Font Registration ---
    console.log("Registering font...");
    const registeredFontName = await registerFontIfNeeded(this.font);
    console.log(`Font registered/retrieved: ${registeredFontName}`);

    // --- Canvas Setup ---
    const cardWidth = 934;
    const cardHeight = 282;
    const avatarSize = 188;
    const padding = 40;
    const avatarX = padding;
    const avatarY = (cardHeight - avatarSize) / 2;
    const borderRadius = 20;
    const statusIndicatorSize = 40;

    console.log("Creating canvas...");
    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");
    console.log("Canvas created.");

    // --- Draw Background & Overlay ---
    console.log("Drawing background...");
    ctx.save();
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, false, false);
    ctx.clip();
    ctx.globalAlpha = 1;
    if (this.background.type === "color") {
      ctx.fillStyle = this.background.value;
      ctx.fillRect(0, 0, cardWidth, cardHeight);
    } else { /* ... image background logic ... */
        try {
            const img = await loadImageWithAxios(this.background.value);
            const aspect = img.width / img.height;
            let drawWidth = cardWidth; let drawHeight = cardWidth / aspect;
            if (drawHeight < cardHeight) { drawHeight = cardHeight; drawWidth = cardHeight * aspect; }
            const offsetX = (cardWidth - drawWidth) / 2; const offsetY = (cardHeight - drawHeight) / 2;
            ctx.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
        } catch (e) {
            console.error("Failed to draw background image:", e.message);
            ctx.fillStyle = "#23272A"; ctx.fillRect(0, 0, cardWidth, cardHeight);
        }
    }
    if (this.overlayOpacity > 0 && this.background.type === "image") {
      ctx.globalAlpha = this.overlayOpacity; ctx.fillStyle = "#000000"; ctx.fillRect(0, 0, cardWidth, cardHeight);
    }
    ctx.restore(); ctx.globalAlpha = 1;
    console.log("Background drawn.");

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      console.log("Drawing card border...");
      ctx.strokeStyle = this.cardBorderColor;
      ctx.lineWidth = 5;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Avatar ---
    console.log("Drawing avatar...");
    ctx.save();
    ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2, 0, Math.PI * 2); ctx.closePath(); ctx.clip();
    try {
      const avatarImg = await loadImageWithAxios(this.avatar);
      ctx.drawImage(avatarImg, avatarX, avatarY, avatarSize, avatarSize);
    } catch (e) { /* ... fallback avatar ... */
        console.error("Failed to draw avatar image:", e.message);
        ctx.fillStyle = "#555"; ctx.fillRect(avatarX, avatarY, avatarSize, avatarSize);
        ctx.fillStyle = "#FFF";
        const fallbackFont = `bold 40px ${registeredFontName}-Bold`;
        console.log(`[Fallback Avatar] Setting font: ${fallbackFont}`);
        ctx.font = fallbackFont;
        ctx.textAlign = "center"; ctx.textBaseline = "middle";
        console.log(`[Fallback Avatar] Filling text: ? at ${avatarX + avatarSize / 2}, ${avatarY + avatarSize / 2}`);
        ctx.fillText("?", avatarX + avatarSize / 2, avatarY + avatarSize / 2);
    }
    ctx.restore();
    console.log("Avatar drawn.");

    // --- Draw Avatar Border (Optional) ---
    if (this.avatarBorderColor) {
        console.log("Drawing avatar border...");
        ctx.strokeStyle = this.avatarBorderColor;
        ctx.lineWidth = 6;
        ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2 + ctx.lineWidth / 2, 0, Math.PI * 2); ctx.stroke(); ctx.closePath();
    }

    // --- Draw Status Indicator ---
    console.log("Drawing status indicator...");
    ctx.fillStyle = this.status.color;
    ctx.beginPath();
    const statusX = avatarX + avatarSize - statusIndicatorSize * 0.8;
    const statusY = avatarY + avatarSize - statusIndicatorSize * 0.8;
    ctx.arc(statusX, statusY, statusIndicatorSize / 2, 0, Math.PI * 2); ctx.fill(); ctx.closePath();
    ctx.strokeStyle = this.background.type === "color" ? this.background.value : "#23272A";
    ctx.lineWidth = 4; ctx.stroke();
    console.log("Status indicator drawn.");

    // --- Draw Text and Progress Bar Area ---
    const textStartX = avatarX + avatarSize + padding;
    const textEndX = cardWidth - padding;
    const textAreaWidth = textEndX - textStartX;

    // --- Draw Username ---
    console.log("Drawing username...");
    ctx.fillStyle = this.usernameColor;
    const usernameFont = `36px ${registeredFontName}-Medium`; // Removed weight prefix as per pureimage standard
    console.log(`Setting username font: ${usernameFont}, fillStyle: ${ctx.fillStyle}`);
    ctx.font = usernameFont;
    ctx.textAlign = "start"; ctx.textBaseline = "top";
    const usernameText = this.username.length > 20 ? this.username.slice(0, 17) + "..." : this.username;
    const usernameY = avatarY + 5;
    console.log(`Filling username text: '${usernameText}' at (${textStartX}, ${usernameY})`);
    ctx.fillText(usernameText, textStartX, usernameY);
    console.log("Username drawn.");

    // --- Draw Rank (Right Aligned) ---
    if (this.rank !== null) {
        console.log("Drawing rank...");
        ctx.fillStyle = this.rankColor;
        const rankFont = `30px ${registeredFontName}-Medium`;
        console.log(`Setting rank font: ${rankFont}, fillStyle: ${ctx.fillStyle}`);
        ctx.font = rankFont;
        ctx.textAlign = "end"; ctx.textBaseline = "top";
        const rankText = `#${formatNumber(this.rank)}`;
        console.log(`Filling rank text: '${rankText}' at (${textEndX}, ${usernameY + 5})`);
        ctx.fillText(rankText, textEndX, usernameY + 5);
        console.log("Rank drawn.");
    }

    // --- Draw Level (Right Aligned, below Rank) ---
    console.log("Drawing level...");
    ctx.fillStyle = this.levelColor;
    const levelFont = `36px ${registeredFontName}-Bold`; // Keep Bold for level
    console.log(`Setting level font: ${levelFont}, fillStyle: ${ctx.fillStyle}`);
    ctx.font = levelFont;
    ctx.textAlign = "end"; ctx.textBaseline = "bottom";
    const levelText = `LEVEL ${formatNumber(this.level)}`;
    const progressBarY = cardHeight - padding - 30;
    const levelY = progressBarY - 15;
    console.log(`Filling level text: '${levelText}' at (${textEndX}, ${levelY})`);
    ctx.fillText(levelText, textEndX, levelY);
    console.log("Level drawn.");

    // --- Draw XP Text (Left Aligned, below Username) ---
    console.log("Drawing XP text...");
    ctx.fillStyle = this.xpColor;
    const xpFont = `24px ${registeredFontName}-Regular`;
    console.log(`Setting XP font: ${xpFont}, fillStyle: ${ctx.fillStyle}`);
    ctx.font = xpFont;
    ctx.textAlign = "start"; ctx.textBaseline = "bottom";
    const xpText = `${formatNumber(this.currentXp)} / ${formatNumber(this.requiredXp)} XP`;
    console.log(`Filling XP text: '${xpText}' at (${textStartX}, ${levelY})`);
    ctx.fillText(xpText, textStartX, levelY);
    console.log("XP text drawn.");

    // --- Draw Progress Bar ---
    console.log("Drawing progress bar...");
    const progressBarHeight = 20;
    const barWidth = textAreaWidth;
    const barRadius = progressBarHeight / 2;

    ctx.fillStyle = this.progressBarBaseColor;
    roundRect(ctx, textStartX, progressBarY, barWidth, progressBarHeight, barRadius, true, false);

    const progressWidth = this._calculateProgress(barWidth);
    if (progressWidth > 0) {
      ctx.fillStyle = this.progressBarColor;
      roundRect(ctx, textStartX, progressBarY, progressWidth, progressBarHeight, barRadius, true, false);
    }
    console.log("Progress bar drawn.");

    // --- Encode and Return Buffer ---
    console.log("Encoding image...");
    try {
      const buffer = await encodeToBuffer(canvas);
      console.log("--- Rank Card Build End ---");
      return buffer;
    } catch (err) {
      console.error("Failed to encode Rank card:", err);
      throw new Error("Could not generate the Rank card image buffer.");
    }
  }
};
