// @cognima/banners/src/server-info-card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    formatNumber,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class ServerInfoCard
 * @classdesc Generates a card displaying server information.
 * @example const card = new ServerInfoCard()
 * .setServerName("Awesome Community")
 * .setServerIcon("server_icon.png")
 * .setMemberCount(12345)
 * .setOnlineCount(5678)
 * .setBoostLevel(3) // Optional
 * .setServerBanner("banner.jpg") // Optional header banner
 * .build();
 */
module.exports = class ServerInfoCard {
  constructor(options) {
    // Core Data
    this.serverName = "Server Name";
    this.serverIcon = null;
    this.serverBanner = null;
    this.memberCount = 0;
    this.onlineCount = 0;
    this.boostLevel = null;
    this.verificationLevel = null;
    this.creationDate = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.bannerBackgroundColor = "#4A5568";
    this.bodyBackgroundColor = "#2C3136";
    this.serverNameColor = "#FFFFFF";
    this.memberCountColor = "#B9BBBE";
    this.onlineCountColor = "#43B581";
    this.boostColor = "#FAA61A";
    this.detailsColor = "#B9BBBE";
    this.iconBorderColor = null;
    this.cardBorderColor = null;
    this.overlayOpacity = 0.3;
  }

  // --- Setters for Core Data ---
  setServerName(name) { if (!name || typeof name !== "string") throw new Error("Server name must be a non-empty string."); this.serverName = name; return this; }
  setServerIcon(image) { if (!image) throw new Error("Server icon source cannot be empty."); this.serverIcon = image; return this; }
  setServerBanner(image) { this.serverBanner = image; return this; }
  setMemberCount(count) { if (typeof count !== "number" || count < 0) throw new Error("Member count must be a non-negative number."); this.memberCount = count; return this; }
  setOnlineCount(count) { if (typeof count !== "number" || count < 0) throw new Error("Online count must be a non-negative number."); this.onlineCount = count; return this; }
  setBoostLevel(level) { if (level !== null && (typeof level !== "number" || level < 0)) throw new Error("Boost level must be a non-negative number or null."); this.boostLevel = level; return this; }
  setVerificationLevel(level) { if (level && typeof level !== "string") throw new Error("Verification level must be a string if provided."); this.verificationLevel = level; return this; }
  setCreationDate(date) { if (date && typeof date !== "string") throw new Error("Creation date must be a string if provided."); this.creationDate = date; return this; }

  // --- Setters for Visual Customization ---
  setBannerBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid banner background color. Use hex format."); this.bannerBackgroundColor = color; return this; }
  setBodyBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid body background color. Use hex format."); this.bodyBackgroundColor = color; return this; }
  setServerNameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid server name color. Use hex format."); this.serverNameColor = color; return this; }
  setMemberCountColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid member count color. Use hex format."); this.memberCountColor = color; return this; }
  setOnlineCountColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid online count color. Use hex format."); this.onlineCountColor = color; return this; }
  setBoostColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid boost color. Use hex format."); this.boostColor = color; return this; }
  setDetailsColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid details color. Use hex format."); this.detailsColor = color; return this; }
  setIconBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid icon border color. Use hex format."); this.iconBorderColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }
  setOverlayOpacity(opacity) { if (typeof opacity !== "number" || opacity < 0 || opacity > 1) throw new Error("Overlay opacity must be between 0 and 1."); this.overlayOpacity = opacity; return this; }

  // --- Build Method ---
  async build() {
    if (!this.serverIcon) throw new Error("Server icon must be set using setServerIcon().");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup ---
    const cardWidth = 600;
    const bannerHeight = 150;
    const iconSize = 100;
    const iconOverlap = 30;
    const bodyPadding = 25;
    const contentStartY = bannerHeight - iconOverlap + iconSize / 2 + bodyPadding;
    let cardHeight = contentStartY + 120; // Initial height
    const borderRadius = 15;

    // Estimate height needed for details
    let detailsHeight = 0;
    if (this.boostLevel !== null) detailsHeight += 30;
    if (this.verificationLevel) detailsHeight += 30;
    if (this.creationDate) detailsHeight += 30;
    cardHeight = contentStartY + 80 + detailsHeight + bodyPadding;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Main Body Background ---
    ctx.fillStyle = this.bodyBackgroundColor;
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, true, false);

    // --- Draw Banner Background & Overlay ---
    ctx.save();
    // Define the clipping path for the banner (top rounded corners only)
    ctx.beginPath();
    ctx.moveTo(0, bannerHeight); // Start bottom-left
    ctx.lineTo(0, borderRadius); // Left edge up to radius
    ctx.quadraticCurveTo(0, 0, borderRadius, 0); // Top-left corner
    ctx.lineTo(cardWidth - borderRadius, 0); // Top edge
    ctx.quadraticCurveTo(cardWidth, 0, cardWidth, borderRadius); // Top-right corner
    ctx.lineTo(cardWidth, bannerHeight); // Right edge down
    ctx.closePath(); // Close path back to bottom-right (implicitly)
    ctx.clip();

    ctx.globalAlpha = 1;
    if (this.serverBanner) {
      try {
        const img = await loadImageWithAxios(this.serverBanner);
        const aspect = img.width / img.height;
        let drawWidth = cardWidth; let drawHeight = cardWidth / aspect;
        if (drawHeight < bannerHeight) { drawHeight = bannerHeight; drawWidth = bannerHeight * aspect; }
        const offsetX = (cardWidth - drawWidth) / 2; const offsetY = 0;
        ctx.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
      } catch (e) {
        console.error("Failed to draw server banner image:", e.message);
        ctx.fillStyle = this.bannerBackgroundColor; ctx.fillRect(0, 0, cardWidth, bannerHeight);
      }
    } else {
      ctx.fillStyle = this.bannerBackgroundColor; ctx.fillRect(0, 0, cardWidth, bannerHeight);
    }

    if (this.overlayOpacity > 0 && this.serverBanner) {
      ctx.globalAlpha = this.overlayOpacity; ctx.fillStyle = "#000000"; ctx.fillRect(0, 0, cardWidth, bannerHeight);
    }
    ctx.restore(); ctx.globalAlpha = 1;

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor; ctx.lineWidth = 4;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Server Icon ---
    const iconX = bodyPadding;
    const iconY = bannerHeight - iconOverlap - iconSize / 2;

    ctx.save();
    ctx.beginPath(); ctx.arc(iconX + iconSize / 2, iconY + iconSize / 2, iconSize / 2, 0, Math.PI * 2); ctx.closePath(); ctx.clip();
    try {
      const iconImg = await loadImageWithAxios(this.serverIcon);
      ctx.drawImage(iconImg, iconX, iconY, iconSize, iconSize);
    } catch (e) { /* ... fallback icon ... */
        console.error("Failed to draw server icon image:", e.message);
        ctx.fillStyle = "#555"; ctx.fillRect(iconX, iconY, iconSize, iconSize);
        ctx.fillStyle = "#FFF";
        const fallbackFont = `24px ${registeredFontName}-Bold`; // Use registered name
        ctx.font = fallbackFont;
        ctx.textAlign = "center"; ctx.textBaseline = "middle";
        ctx.fillText("?", iconX + iconSize / 2, iconY + iconSize / 2);
    }
    ctx.restore();

    // --- Draw Icon Border (Optional) ---
    if (this.iconBorderColor) {
        ctx.strokeStyle = this.iconBorderColor; ctx.lineWidth = 4;
        ctx.beginPath(); ctx.arc(iconX + iconSize / 2, iconY + iconSize / 2, iconSize / 2 + ctx.lineWidth / 2, 0, Math.PI * 2); ctx.stroke(); ctx.closePath();
    }

    // --- Draw Server Name ---
    ctx.fillStyle = this.serverNameColor;
    const serverNameFont = `26px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = serverNameFont;
    ctx.textAlign = "start"; ctx.textBaseline = "top";
    const nameX = iconX + iconSize + bodyPadding;
    const nameY = bannerHeight + 10;
    const nameText = this.serverName.length > 25 ? this.serverName.slice(0, 22) + "..." : this.serverName;
    ctx.fillText(nameText, nameX, nameY);

    // --- Draw Content Below Icon ---
    let currentY = contentStartY;
    const contentX = bodyPadding;
    const contentWidth = cardWidth - 2 * bodyPadding;

    // Member Counts
    ctx.textAlign = "start"; ctx.textBaseline = "top";
    const memberCountFont = `18px ${registeredFontName}-Medium`; // Use registered name
    ctx.font = memberCountFont;

    // Online Count
    ctx.fillStyle = this.onlineCountColor;
    ctx.beginPath(); ctx.arc(contentX + 8, currentY + 9, 6, 0, Math.PI * 2); ctx.fill(); ctx.closePath();
    ctx.fillStyle = this.memberCountColor;
    ctx.fillText(`${formatNumber(this.onlineCount)} Online`, contentX + 20, currentY);

    // Total Members
    const onlineTextWidth = ctx.measureText(`${formatNumber(this.onlineCount)} Online`).width;
    const totalMembersX = contentX + 20 + onlineTextWidth + 25;
    ctx.fillStyle = "#8E9297";
    ctx.beginPath(); ctx.arc(totalMembersX + 8, currentY + 9, 6, 0, Math.PI * 2); ctx.fill(); ctx.closePath();
    ctx.fillStyle = this.memberCountColor;
    ctx.fillText(`${formatNumber(this.memberCount)} Members`, totalMembersX + 20, currentY);

    currentY += 40;

    // Optional Details
    const detailsFont = `16px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = detailsFont;
    ctx.fillStyle = this.detailsColor;

    if (this.boostLevel !== null) {
        ctx.fillStyle = this.boostColor;
        ctx.fillText(`🚀 Boost Level ${this.boostLevel}`, contentX, currentY);
        currentY += 25;
        ctx.fillStyle = this.detailsColor;
    }
    if (this.verificationLevel) {
        ctx.fillText(`✅ Verification: ${this.verificationLevel}`, contentX, currentY);
        currentY += 25;
    }
    if (this.creationDate) {
        ctx.fillText(`📅 Created: ${this.creationDate}`, contentX, currentY);
        currentY += 25;
    }

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Server Info Card:", err);
      throw new Error("Could not generate the Server Info Card image buffer.");
    }
  }
};
