// @cognima/banners/src/spotify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect, // Keep using roundRect for general shapes
    formatTime,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class Spotify
 * @classdesc Generates a Spotify "Now Playing" card with artwork, song details, and progress bar.
 * @example const card = new Spotify()
 * .setAuthor("Artist Name")
 * .setAlbum("Album Name")
 * .setStartTimestamp(60000) // 1 minute in ms
 * .setEndTimestamp(240000) // 4 minutes in ms
 * .setImage("image_url.png")
 * .setTitle("Song Title")
 * .build();
 */
module.exports = class Spotify {
  constructor(options) {
    // Core Data
    this.author = null;
    this.album = null;
    this.start = null;
    this.end = null;
    this.image = null;
    this.title = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.backgroundColor = "#1ED760"; // Spotify Green
    this.imagePlaceholderColor = "#191414"; // Spotify Dark Grey
    this.authorColor = "#FFFFFF";
    this.titleColor = "#FFFFFF";
    this.albumColor = "#FFFFFF";
    this.timestampColor = "#FFFFFF";
    this.progressBarColor = "#FFFFFF";
    this.progressBarBaseColor = "#B3B3B3"; // Lighter grey for base
    this.cardBorderColor = null;
  }

  // --- Setters for Core Data ---
  setAuthor(name) { if (!name || typeof name !== "string") throw new Error("Author name must be a non-empty string."); this.author = name; return this; }
  setAlbum(name) { if (!name || typeof name !== "string") throw new Error("Album name must be a non-empty string."); this.album = name; return this; }
  setStartTimestamp(ms) { if (typeof ms !== "number" || ms < 0) throw new Error("Start timestamp must be a non-negative number (milliseconds)."); this.start = ms; return this; }
  setEndTimestamp(ms) { if (typeof ms !== "number" || ms <= 0) throw new Error("End timestamp must be a positive number (milliseconds)."); this.end = ms; return this; }
  setImage(image) { if (!image) throw new Error("Image source cannot be empty."); this.image = image; return this; }
  setTitle(title) { if (!title || typeof title !== "string") throw new Error("Title must be a non-empty string."); this.title = title; return this; }

  // --- Setters for Visual Customization ---
  setBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid background color. Use hex format."); this.backgroundColor = color; return this; }
  setImagePlaceholderColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid image placeholder color. Use hex format."); this.imagePlaceholderColor = color; return this; }
  setAuthorColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid author color. Use hex format."); this.authorColor = color; return this; }
  setTitleColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid title color. Use hex format."); this.titleColor = color; return this; }
  setAlbumColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid album color. Use hex format."); this.albumColor = color; return this; }
  setTimestampColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid timestamp color. Use hex format."); this.timestampColor = color; return this; }
  setProgressBarColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar color. Use hex format."); this.progressBarColor = color; return this; }
  setProgressBarBaseColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar base color. Use hex format."); this.progressBarBaseColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }

  // --- Private Helper Methods ---
  _calculateProgress(barWidth) {
    if (this.end <= 0) return 0;
    const progressRatio = Math.max(0, Math.min(this.start / this.end, 1));
    return progressRatio * barWidth;
  }

  // --- Build Method ---
  async build() {
    // --- Validation ---
    if (!this.author) throw new Error("Author must be set using setAuthor().");
    if (!this.album) throw new Error("Album must be set using setAlbum().");
    if (this.start === null) throw new Error("Start timestamp must be set using setStartTimestamp().");
    if (this.end === null) throw new Error("End timestamp must be set using setEndTimestamp().");
    if (!this.image) throw new Error("Image must be set using setImage().");
    if (!this.title) throw new Error("Title must be set using setTitle().");
    if (this.start > this.end) console.warn("Start timestamp is greater than end timestamp.");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup ---
    const cardWidth = 900;
    const cardHeight = 280;
    const imageSize = cardHeight; // Image takes full height
    const borderRadius = 15;
    const padding = 30;
    const textPadding = 25;
    const progressBarHeight = 8;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Background ---
    ctx.fillStyle = this.backgroundColor;
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, true, false);

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor;
      ctx.lineWidth = 4;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Image ---
    const imageX = 0;
    const imageY = 0;
    const imageRadius = borderRadius; // Match card radius for seamless edge

    // Define the clipping path for the left side with rounded corners
    const clipPath = () => {
        ctx.beginPath();
        ctx.moveTo(imageX + imageRadius, imageY); // Top edge start
        ctx.lineTo(imageX + imageSize, imageY); // Top edge end
        ctx.lineTo(imageX + imageSize, imageY + cardHeight); // Right edge
        ctx.lineTo(imageX + imageRadius, imageY + cardHeight); // Bottom edge start
        ctx.quadraticCurveTo(imageX, imageY + cardHeight, imageX, imageY + cardHeight - imageRadius); // Bottom-left corner
        ctx.lineTo(imageX, imageY + imageRadius); // Left edge
        ctx.quadraticCurveTo(imageX, imageY, imageX + imageRadius, imageY); // Top-left corner
        ctx.closePath();
    };

    try {
      const img = await loadImageWithAxios(this.image);
      ctx.save();
      clipPath(); // Apply the clipping path
      ctx.clip();
      // Draw image scaled to fit height
      const aspect = img.width / img.height;
      let drawWidth = imageSize * aspect;
      let drawHeight = imageSize;
      // Center the image if aspect ratio doesn\t match square
      let imgDrawX = imageX;
      if (drawWidth > imageSize) {
          imgDrawX = imageX - (drawWidth - imageSize) / 2;
      }
      ctx.drawImage(img, imgDrawX, imageY, drawWidth, drawHeight);
      ctx.restore();
    } catch (e) {
      console.error("Failed to draw artwork image:", e.message);
      ctx.save();
      clipPath(); // Apply the clipping path for placeholder too
      ctx.clip();
      ctx.fillStyle = this.imagePlaceholderColor;
      ctx.fillRect(imageX, imageY, imageSize, cardHeight); // Fill the clipped area
      ctx.restore();
    }

    // --- Text and Progress Bar Area ---
    const textStartX = imageSize + textPadding;
    const textEndX = cardWidth - padding;
    const textWidth = textEndX - textStartX;
    let currentY = padding + 10;

    // --- Draw Text ---
    ctx.textAlign = "start";
    ctx.textBaseline = "top";

    // Title (Bold, large)
    ctx.fillStyle = this.titleColor;
    const titleFont = `34px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = titleFont;
    const titleText = this.title.length > 30 ? this.title.slice(0, 27) + "..." : this.title;
    ctx.fillText(titleText, textStartX, currentY);
    currentY += 45;

    // Author (Medium, medium)
    ctx.fillStyle = this.authorColor;
    const authorFont = `28px ${registeredFontName}-Medium`; // Use registered name
    ctx.font = authorFont;
    const authorText = this.author.length > 35 ? this.author.slice(0, 32) + "..." : this.author;
    ctx.fillText(authorText, textStartX, currentY);
    currentY += 38;

    // Album (Regular, smaller)
    ctx.fillStyle = this.albumColor;
    const albumFont = `24px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = albumFont;
    const albumText = this.album.length > 40 ? this.album.slice(0, 37) + "..." : this.album;
    ctx.fillText(albumText, textStartX, currentY);
    currentY += 35;

    // --- Draw Progress Bar & Timestamps ---
    const barBaseY = cardHeight - padding - progressBarHeight - 15; // Position near bottom
    const barWidth = textWidth;
    const barStartX = textStartX;
    const barRadius = progressBarHeight / 2;
    const timestampY = barBaseY + progressBarHeight + 8; // Below bar

    // Bar background
    ctx.fillStyle = this.progressBarBaseColor;
    roundRect(ctx, barStartX, barBaseY, barWidth, progressBarHeight, barRadius, true, false);

    // Bar progress
    const progressWidth = this._calculateProgress(barWidth);
    if (progressWidth > 0) {
      ctx.fillStyle = this.progressBarColor;
      roundRect(ctx, barStartX, barBaseY, progressWidth, progressBarHeight, barRadius, true, false);
    }

    // Timestamps
    const start_format = formatTime(this.start);
    const end_format = formatTime(this.end);
    ctx.fillStyle = this.timestampColor;
    const timestampFont = `20px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = timestampFont;
    ctx.textBaseline = "top";
    ctx.fillText(start_format, barStartX, timestampY);
    ctx.textAlign = "end";
    ctx.fillText(end_format, barStartX + barWidth, timestampY);

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Spotify card:", err);
      throw new Error("Could not generate the Spotify card image buffer.");
    }
  }
};
