// @cognima/banners/src/spotify-minimal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect, // Keep using roundRect for general shapes
    formatTime,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class SpotifyMinimal
 * @classdesc Generates a minimal Spotify "Now Playing" card, focusing on artwork and basic info.
 * @example const card = new SpotifyMinimal()
 * .setAuthor("Artist Name")
 * .setTitle("Song Title")
 * .setImage("image_url.png")
 * .setTimestamp(60000, 180000) // 1:00 / 3:00
 * .build();
 */
module.exports = class SpotifyMinimal {
  constructor(options) {
    // Core Data
    this.author = null;
    this.title = null;
    this.image = null;
    this.start = null;
    this.end = null;

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.backgroundColor = "#191414";
    this.imagePlaceholderColor = "#282828";
    this.textColor = "#FFFFFF";
    this.progressBarColor = "#1DB954";
    this.progressBarBaseColor = "#535353";
    this.cardBorderColor = null;
  }

  // --- Setters for Core Data ---
  setAuthor(name) { if (!name || typeof name !== "string") throw new Error("Author name must be a non-empty string."); this.author = name; return this; }
  setTitle(title) { if (!title || typeof title !== "string") throw new Error("Title must be a non-empty string."); this.title = title; return this; }
  setImage(image) { if (!image) throw new Error("Image source cannot be empty."); this.image = image; return this; }
  setTimestamp(start, end) { if (typeof start !== "number" || start < 0) throw new Error("Start timestamp must be a non-negative number."); if (typeof end !== "number" || end <= 0) throw new Error("End timestamp must be a positive number."); if (start > end) console.warn("Start timestamp is greater than end timestamp."); this.start = start; this.end = end; return this; }

  // --- Setters for Visual Customization ---
  setBackgroundColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid background color. Use hex format."); this.backgroundColor = color; return this; }
  setImagePlaceholderColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid image placeholder color. Use hex format."); this.imagePlaceholderColor = color; return this; }
  setTextColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid text color. Use hex format."); this.textColor = color; return this; }
  setProgressBarColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar color. Use hex format."); this.progressBarColor = color; return this; }
  setProgressBarBaseColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid progress bar base color. Use hex format."); this.progressBarBaseColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }

  // --- Private Helper Methods ---
  _calculateProgress(barWidth) {
    if (this.end <= 0) return 0;
    const progressRatio = Math.max(0, Math.min(this.start / this.end, 1));
    return progressRatio * barWidth;
  }

  // --- Build Method ---
  async build() {
    // --- Validation ---
    if (!this.author) throw new Error("Author must be set using setAuthor().");
    if (!this.title) throw new Error("Title must be set using setTitle().");
    if (!this.image) throw new Error("Image must be set using setImage().");
    if (this.start === null || this.end === null) throw new Error("Timestamps must be set using setTimestamp().");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup (Minimalist Dimensions) ---
    const cardWidth = 450;
    const cardHeight = 150;
    const imageSize = cardHeight;
    const borderRadius = 10;
    const padding = 0;
    const textPadding = 15;
    const progressBarHeight = 4;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Background ---
    ctx.fillStyle = this.backgroundColor;
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, true, false);

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor;
      ctx.lineWidth = 3;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Image ---
    const imageX = 0;
    const imageY = 0;
    const imageRadius = borderRadius;

    // Define the clipping path for the left side with rounded corners
    const clipPath = () => {
        ctx.beginPath();
        ctx.moveTo(imageX + imageRadius, imageY); // Top edge start
        ctx.lineTo(imageX + imageSize, imageY); // Top edge end
        ctx.lineTo(imageX + imageSize, imageY + cardHeight); // Right edge
        ctx.lineTo(imageX + imageRadius, imageY + cardHeight); // Bottom edge start
        ctx.quadraticCurveTo(imageX, imageY + cardHeight, imageX, imageY + cardHeight - imageRadius); // Bottom-left corner
        ctx.lineTo(imageX, imageY + imageRadius); // Left edge
        ctx.quadraticCurveTo(imageX, imageY, imageX + imageRadius, imageY); // Top-left corner
        ctx.closePath();
    };

    try {
      const img = await loadImageWithAxios(this.image);
      ctx.save();
      clipPath(); // Apply the clipping path
      ctx.clip();
      // Draw image scaled to fit height
      const aspect = img.width / img.height;
      let drawWidth = imageSize * aspect;
      let drawHeight = imageSize;
      // Center the image if aspect ratio doesn\t match square
      let imgDrawX = imageX;
      if (drawWidth > imageSize) {
          imgDrawX = imageX - (drawWidth - imageSize) / 2;
      }
      ctx.drawImage(img, imgDrawX, imageY, drawWidth, drawHeight);
      ctx.restore();
    } catch (e) {
      console.error("Failed to draw artwork image:", e.message);
      ctx.save();
      clipPath(); // Apply the clipping path for placeholder too
      ctx.clip();
      ctx.fillStyle = this.imagePlaceholderColor;
      ctx.fillRect(imageX, imageY, imageSize, cardHeight); // Fill the clipped area
      ctx.restore();
    }

    // --- Text and Progress Bar Area ---
    const textStartX = imageSize + textPadding;
    const textEndX = cardWidth - textPadding;
    const textWidth = textEndX - textStartX;

    // --- Draw Text ---
    ctx.textAlign = "start";
    ctx.textBaseline = "middle";
    ctx.fillStyle = this.textColor;

    // Title (Bold, larger)
    const titleFont = `18px ${registeredFontName}-Bold`; // Use registered name
    ctx.font = titleFont;
    const titleText = this.title.length > 25 ? this.title.slice(0, 22) + "..." : this.title;
    const titleY = cardHeight * 0.35;
    ctx.fillText(titleText, textStartX, titleY);

    // Author (Regular, smaller)
    const authorFont = `15px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = authorFont;
    const authorText = this.author.length > 30 ? this.author.slice(0, 27) + "..." : this.author;
    const authorY = cardHeight * 0.60;
    ctx.fillText(authorText, textStartX, authorY);

    // --- Draw Progress Bar & Timestamps ---
    const barBaseY = cardHeight - textPadding - progressBarHeight;
    const barWidth = textWidth;
    const barStartX = textStartX;
    const barRadius = progressBarHeight / 2;
    const timestampY = barBaseY - 12;

    // Timestamps
    const start_format = formatTime(this.start);
    const end_format = formatTime(this.end);
    const timestampFont = `12px ${registeredFontName}-Regular`; // Use registered name
    ctx.font = timestampFont;
    ctx.textBaseline = "bottom";
    ctx.fillText(start_format, barStartX, timestampY);
    ctx.textAlign = "end";
    ctx.fillText(end_format, barStartX + barWidth, timestampY);

    // Bar background
    ctx.fillStyle = this.progressBarBaseColor;
    roundRect(ctx, barStartX, barBaseY, barWidth, progressBarHeight, barRadius, true, false);

    // Bar progress
    const progressWidth = this._calculateProgress(barWidth);
    if (progressWidth > 0) {
      ctx.fillStyle = this.progressBarColor;
      roundRect(ctx, barStartX, barBaseY, progressWidth, progressBarHeight, barRadius, true, false);
    }

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Spotify Minimal card:", err);
      throw new Error("Could not generate the Spotify Minimal card image buffer.");
    }
  }
};
