// @cognima/banners/src/utils.js
"use strict";

const pureimage = require("pureimage");
const fs = require("fs");
const path = require("path");
const axios = require("axios");
const { Readable } = require("stream");

// --- Constants for Standardization ---
const DEFAULT_FONT_FAMILY = "Poppins";
const DEFAULT_FONT_DIR = path.join(__dirname, "../assets/fonts/Poppins");
const FONT_WEIGHTS = {
    bold: "Poppins-Bold.ttf",
    medium: "Poppins-Medium.ttf",
    regular: "Poppins-Regular.ttf",
};

// --- Color Validation ---
function isValidHexColor(color) {
    return /^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$/.test(color);
}

// --- Font Registration ---
let fontsRegistered = false;
const registeredFontNames = new Set(); // Keep track of successfully registered names

async function registerDefaultFonts() {
    if (fontsRegistered) {
        // console.log("Default fonts already registered. Using:", DEFAULT_FONT_FAMILY);
        return DEFAULT_FONT_FAMILY;
    }
    console.log("Attempting to register default fonts...");
    try {
        const fontPromises = [];
        for (const weight in FONT_WEIGHTS) {
            const fontPath = path.join(DEFAULT_FONT_DIR, FONT_WEIGHTS[weight]);
            if (fs.existsSync(fontPath)) {
                const fontName = `${DEFAULT_FONT_FAMILY}-${weight.charAt(0).toUpperCase() + weight.slice(1)}`;
                console.log(`Registering font: ${fontName} from ${fontPath}`);
                const font = pureimage.registerFont(fontPath, fontName);
                fontPromises.push(
                    font.load().then(() => {
                        console.log(`Successfully loaded and registered: ${fontName}`);
                        registeredFontNames.add(fontName);
                    }).catch(err => {
                        console.error(`Failed to load font ${fontName}:`, err);
                        throw err; // Propagate error
                    })
                );
            } else {
                console.warn(`Default font file not found: ${fontPath}`);
            }
        }
        if (fontPromises.length > 0) {
            await Promise.all(fontPromises);
            if (registeredFontNames.size === Object.keys(FONT_WEIGHTS).length) {
                 fontsRegistered = true;
                 console.log("All default fonts registered successfully. Base name:", DEFAULT_FONT_FAMILY);
                 return DEFAULT_FONT_FAMILY;
            } else {
                throw new Error(`Only ${registeredFontNames.size}/${Object.keys(FONT_WEIGHTS).length} default fonts registered successfully.`);
            }
        } else {
            throw new Error("No default font files found in expected directory.");
        }
    } catch (err) {
        console.error("Error registering default fonts:", err);
        throw new Error("Failed to load default fonts. Ensure Poppins font files (Bold, Medium, Regular) are in assets/fonts/Poppins.");
    }
}

async function registerFontIfNeeded(fontConfig) {
    if (!fontConfig || !fontConfig.path || !fontConfig.name) {
        // Use default fonts if no custom font provided or config is incomplete
        // console.log("No custom font config, using default fonts.");
        return await registerDefaultFonts();
    }

    console.log(`Attempting to register custom font: ${fontConfig.name} from ${fontConfig.path}`);
    try {
        if (fs.existsSync(fontConfig.path)) {
            const font = pureimage.registerFont(fontConfig.path, fontConfig.name);
            await font.load();
            console.log(`Successfully loaded and registered custom font: ${fontConfig.name}`);
            registeredFontNames.add(fontConfig.name);
            return fontConfig.name; // Return the name of the registered custom font
        } else {
            console.warn(`Custom font file not found: ${fontConfig.path}. Falling back to default fonts.`);
            return await registerDefaultFonts();
        }
    } catch (err) {
        console.error(`Error registering custom font ${fontConfig.name}:`, err);
        console.warn("Falling back to default fonts.");
        return await registerDefaultFonts();
    }
}

// --- Image Loading ---
async function loadImageWithAxios(source) {
    if (!source) throw new Error("Image source cannot be null or empty.");

    if (typeof source === "string" && source.startsWith("http")) {
        try {
            const response = await axios.get(source, { responseType: "arraybuffer" });
            const buffer = Buffer.from(response.data, "binary");
            const contentType = response.headers["content-type"];

            if (contentType && contentType.includes("png")) {
                return await pureimage.decodePNGFromStream(Readable.from(buffer));
            } else if (contentType && (contentType.includes("jpeg") || contentType.includes("jpg"))) {
                return await pureimage.decodeJPEGFromStream(Readable.from(buffer));
            } else {
                // Attempt decoding, prioritizing PNG then JPEG
                try { return await pureimage.decodePNGFromStream(Readable.from(buffer)); } catch (e) { /* ignore */ }
                try { return await pureimage.decodeJPEGFromStream(Readable.from(buffer)); } catch (e) { /* ignore */ }
                throw new Error(`Unsupported image type \'${contentType}\' from URL: ${source}`);
            }
        } catch (error) {
            console.error(`Error loading image from URL ${source}:`, error.message);
            throw new Error(`Failed to load image from URL: ${source}. Ensure the URL is correct and the image format is PNG or JPEG.`);
        }
    } else if (Buffer.isBuffer(source)) {
        try { return await pureimage.decodePNGFromStream(Readable.from(source)); } catch (e) { /* ignore */ }
        try { return await pureimage.decodeJPEGFromStream(Readable.from(source)); } catch (e) { /* ignore */ }
        throw new Error("Failed to decode image from buffer. Ensure it is a valid PNG or JPEG buffer.");
    } else if (typeof source === "string") {
        // Assume local path
        try {
            const stream = fs.createReadStream(source);
            if (source.toLowerCase().endsWith(".png")) return await pureimage.decodePNGFromStream(stream);
            if (source.toLowerCase().endsWith(".jpg") || source.toLowerCase().endsWith(".jpeg")) return await pureimage.decodeJPEGFromStream(stream);
            throw new Error("Unsupported local image file type. Only PNG and JPEG are supported.");
        } catch (error) {
            console.error(`Error loading image from path ${source}:`, error.message);
            throw new Error(`Failed to load image from path: ${source}. Ensure the path is correct and the file is a PNG or JPEG.`);
        }
    } else if (source && source.width && source.height && source.data) {
        return source; // Already a bitmap
    }
    throw new Error("Invalid image source provided. Must be a URL (http/https), Buffer, local path, or a pureimage bitmap.");
}

// --- Canvas Encoding ---
async function encodeToBuffer(canvas) {
    const writableStream = new (require("stream").Writable)({
        write(chunk, encoding, callback) {
            if (!this.data) this.data = [];
            this.data.push(chunk);
            callback();
        },
    });
    await pureimage.encodePNGToStream(canvas, writableStream);
    if (!writableStream.data) throw new Error("Failed to encode image to buffer.");
    return Buffer.concat(writableStream.data);
}

// --- Drawing Helpers ---
function roundRect(ctx, x, y, width, height, radius, fill, stroke) {
    if (typeof radius === "number") {
        radius = { tl: radius, tr: radius, br: radius, bl: radius };
    } else {
        radius = { ...{ tl: 0, tr: 0, br: 0, bl: 0 }, ...radius };
    }
    ctx.beginPath();
    ctx.moveTo(x + radius.tl, y);
    ctx.lineTo(x + width - radius.tr, y);
    ctx.quadraticCurveTo(x + width, y, x + width, y + radius.tr);
    ctx.lineTo(x + width, y + height - radius.br);
    ctx.quadraticCurveTo(x + width, y + height, x + width - radius.br, y + height);
    ctx.lineTo(x + radius.bl, y + height);
    ctx.quadraticCurveTo(x, y + height, x, y + height - radius.bl);
    ctx.lineTo(x, y + radius.tl);
    ctx.quadraticCurveTo(x, y, x + radius.tl, y);
    ctx.closePath();
    if (fill) {
        ctx.fill();
    }
    if (stroke) {
        ctx.stroke();
    }
}

function wrapText(context, text, x, y, maxWidth, lineHeight, baseFontName) {
  const words = text.split(" ");
  let line = "";
  let currentY = y;
  const initialFont = context.font || `16px ${baseFontName}-Regular`; // Fallback if context.font is unset
  const fontMatch = initialFont.match(/\d+px/);
  const fontSize = fontMatch ? fontMatch[0] : "16px"; // Default to 16px if match fails
  const fontWeight = initialFont.includes("Bold") ? "Bold" : initialFont.includes("Medium") ? "Medium" : "Regular";
  const fontName = `${baseFontName}-${fontWeight}`;

  // Log for debugging
  console.log(`wrapText: Initial font='${initialFont}', Base='${baseFontName}', Size='${fontSize}', Weight='${fontWeight}', Final Name='${fontName}'`);

  if (!registeredFontNames.has(fontName)) {
    console.warn(`wrapText: Font '${fontName}' not found in registered fonts. Using fallback font.`);
    context.font = `${fontSize} ${baseFontName}-Regular`; // Fallback to a safe font
  } else {
    context.font = `${fontSize} ${fontName}`; // Set the font explicitly
  }

  for (let n = 0; n < words.length; n++) {
    const testLine = line + words[n] + " ";
    context.font = `${fontSize} ${fontName}`; // Ensure font is set before measuring
    const metrics = context.measureText(testLine);
    const testWidth = metrics.width;

    if (testWidth > maxWidth && n > 0) {
      context.fillText(line.trim(), x, currentY);
      line = words[n] + " ";
      currentY += lineHeight;
    } else {
      line = testLine;
    }
  }
  context.font = `${fontSize} ${fontName}`; // Ensure font is set for the last line
  context.fillText(line.trim(), x, currentY);
  return currentY + lineHeight; // Return the Y position for the next line
}

// --- Formatting Helpers ---
function formatTime(ms) {
    if (typeof ms !== "number" || isNaN(ms) || ms < 0) return "0:00";
    const totalSeconds = Math.floor(ms / 1000);
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = totalSeconds % 60;
    return `${minutes}:${seconds < 10 ? "0" : ""}${seconds}`;
}

function formatNumber(num) {
    if (typeof num !== "number" || isNaN(num)) return "0";
    // Simple K/M formatting for large numbers (optional, adjust thresholds)
    if (num >= 1000000) return (num / 1000000).toFixed(1).replace(/\.0$/, "") + "M";
    if (num >= 1000) return (num / 1000).toFixed(1).replace(/\.0$/, "") + "K";
    return num.toString();
}

// --- Exports ---
module.exports = {
    isValidHexColor,
    registerDefaultFonts,
    registerFontIfNeeded,
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    wrapText,
    formatTime,
    formatNumber,
    DEFAULT_FONT_FAMILY // Export default font family name
};
