// @cognima/banners/src/welcome-leave-alt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const pureimage = require("pureimage");
const path = require("path");
const {
    loadImageWithAxios,
    encodeToBuffer,
    roundRect,
    wrapText,
    registerFontIfNeeded,
    isValidHexColor,
    DEFAULT_FONT_FAMILY
} = require("./utils");

/**
 * @class WelcomeLeaveAlt
 * @classdesc Generates an alternative, more detailed welcome/leave banner with side avatar.
 * @example const card = new WelcomeLeaveAlt()
 * .setAvatar("avatar.png")
 * .setTitle("Welcome")
 * .setUsername("NewUser#1234")
 * .setMessage("Welcome to our awesome server! Enjoy your stay.")
 * .setMemberCount(1234)
 * .build();
 */
module.exports = class WelcomeLeaveAlt {
  constructor(options) {
    // Core Data
    this.avatar = null;
    this.title = "Welcome";
    this.username = "Username#0000";
    this.message = "Message text";
    this.memberCount = null; // Optional member count

    // Visual Customization
    this.font = { name: options?.font?.name ?? DEFAULT_FONT_FAMILY, path: options?.font?.path };
    this.background = { type: "color", value: "#2F3136" };
    this.titleColor = "#FFFFFF";
    this.usernameColor = "#FFFFFF";
    this.messageColor = "#B9BBBE";
    this.memberCountColor = "#B9BBBE";
    this.avatarBorderColor = null;
    this.cardBorderColor = null;
    this.overlayOpacity = 0.4;
  }

  // --- Setters for Core Data ---
  setAvatar(image) { if (!image) throw new Error("Avatar image source cannot be empty."); this.avatar = image; return this; }
  setTitle(text) { if (!text || typeof text !== "string") throw new Error("Title text must be a non-empty string."); this.title = text; return this; }
  setUsername(name) { if (!name || typeof name !== "string") throw new Error("Username must be a non-empty string."); this.username = name; return this; }
  setMessage(text) { if (!text || typeof text !== "string") throw new Error("Message text must be a non-empty string."); this.message = text; return this; }
  setMemberCount(count) { if (count !== null && (typeof count !== "number" || count < 0)) throw new Error("Member count must be a non-negative number or null."); this.memberCount = count; return this; }

  // --- Setters for Visual Customization ---
  setBackground(type, value) {
    const types = ["color", "image"];
    if (!type || !types.includes(type.toLowerCase())) throw new Error("Background type must be \'color\' or \'image\'.");
    if (!value) throw new Error("Background value cannot be empty.");
    if (type.toLowerCase() === "color" && !isValidHexColor(value)) throw new Error("Invalid background color. Use hex format.");
    this.background = { type: type.toLowerCase(), value };
    return this;
  }

  setTitleColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid title color. Use hex format."); this.titleColor = color; return this; }
  setUsernameColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid username color. Use hex format."); this.usernameColor = color; return this; }
  setMessageColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid message color. Use hex format."); this.messageColor = color; return this; }
  setMemberCountColor(color) { if (!color || !isValidHexColor(color)) throw new Error("Invalid member count color. Use hex format."); this.memberCountColor = color; return this; }
  setAvatarBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid avatar border color. Use hex format."); this.avatarBorderColor = color; return this; }
  setCardBorderColor(color) { if (color && !isValidHexColor(color)) throw new Error("Invalid card border color. Use hex format."); this.cardBorderColor = color; return this; }
  setOverlayOpacity(opacity) { if (typeof opacity !== "number" || opacity < 0 || opacity > 1) throw new Error("Overlay opacity must be between 0 and 1."); this.overlayOpacity = opacity; return this; }

  // --- Build Method ---
  async build() {
    if (!this.avatar) throw new Error("Avatar image must be set using setAvatar().");

    // --- Font Registration ---
    const registeredFontName = await registerFontIfNeeded(this.font);

    // --- Canvas Setup ---
    const cardWidth = 800;
    const cardHeight = 250;
    const avatarSize = 160;
    const borderRadius = 20;
    const padding = 30;
    const avatarX = padding;
    const avatarY = (cardHeight - avatarSize) / 2;

    const canvas = pureimage.make(cardWidth, cardHeight);
    const ctx = canvas.getContext("2d");

    // --- Draw Background & Overlay ---
    ctx.save();
    roundRect(ctx, 0, 0, cardWidth, cardHeight, borderRadius, false, false);
    ctx.clip();
    ctx.globalAlpha = 1;
    if (this.background.type === "color") {
      ctx.fillStyle = this.background.value;
      ctx.fillRect(0, 0, cardWidth, cardHeight);
    } else { /* ... image background logic ... */
        try {
            const img = await loadImageWithAxios(this.background.value);
            const aspect = img.width / img.height;
            let drawWidth = cardWidth; let drawHeight = cardWidth / aspect;
            if (drawHeight < cardHeight) { drawHeight = cardHeight; drawWidth = cardHeight * aspect; }
            const offsetX = (cardWidth - drawWidth) / 2; const offsetY = (cardHeight - drawHeight) / 2;
            ctx.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
        } catch (e) {
            console.error("Failed to draw background image:", e.message);
            ctx.fillStyle = "#2F3136"; ctx.fillRect(0, 0, cardWidth, cardHeight);
        }
    }
    if (this.overlayOpacity > 0 && this.background.type === "image") {
      ctx.globalAlpha = this.overlayOpacity; ctx.fillStyle = "#000000"; ctx.fillRect(0, 0, cardWidth, cardHeight);
    }
    ctx.restore(); ctx.globalAlpha = 1;

    // --- Draw Card Border (Optional) ---
    if (this.cardBorderColor) {
      ctx.strokeStyle = this.cardBorderColor;
      ctx.lineWidth = 5;
      roundRect(ctx, ctx.lineWidth / 2, ctx.lineWidth / 2, cardWidth - ctx.lineWidth, cardHeight - ctx.lineWidth, borderRadius - ctx.lineWidth / 2, false, true);
    }

    // --- Draw Avatar ---
    ctx.save();
    ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2, 0, Math.PI * 2); ctx.closePath(); ctx.clip();
    try {
      const avatarImg = await loadImageWithAxios(this.avatar);
      ctx.drawImage(avatarImg, avatarX, avatarY, avatarSize, avatarSize);
    } catch (e) { /* ... fallback avatar ... */
        console.error("Failed to draw avatar image:", e.message);
        ctx.fillStyle = "#555"; ctx.fillRect(avatarX, avatarY, avatarSize, avatarSize);
        ctx.fillStyle = "#FFF"; ctx.font = `bold 40px ${registeredFontName}-Bold`; ctx.textAlign = "center"; ctx.textBaseline = "middle";
        ctx.fillText("?", avatarX + avatarSize / 2, avatarY + avatarSize / 2);
    }
    ctx.restore();

    // --- Draw Avatar Border (Optional) ---
    if (this.avatarBorderColor) {
        ctx.strokeStyle = this.avatarBorderColor;
        ctx.lineWidth = 5;
        ctx.beginPath(); ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2 + ctx.lineWidth / 2, 0, Math.PI * 2); ctx.stroke(); ctx.closePath();
    }

    // --- Draw Text --- (Right of avatar)
    const textStartX = avatarX + avatarSize + padding;
    const textMaxWidth = cardWidth - textStartX - padding;
    let currentY = padding + 10;

    // Title
    ctx.fillStyle = this.titleColor;
    ctx.font = `bold 36px ${registeredFontName}-Bold`;
    ctx.textAlign = "start"; ctx.textBaseline = "top";
    const titleText = this.title.length > 25 ? this.title.slice(0, 22) + "..." : this.title;
    ctx.fillText(titleText, textStartX, currentY);
    currentY += 45;

    // Username
    ctx.fillStyle = this.usernameColor;
    ctx.font = `medium 30px ${registeredFontName}-Medium`;
    const usernameText = this.username.length > 30 ? this.username.slice(0, 27) + "..." : this.username;
    ctx.fillText(usernameText, textStartX, currentY);
    currentY += 40;

    // Message
    ctx.fillStyle = this.messageColor;
    ctx.font = `regular 24px ${registeredFontName}-Regular`;
    currentY = wrapText(ctx, this.message, textStartX, currentY, textMaxWidth, 30, registeredFontName);
    currentY += 5; // Small gap before member count

    // Member Count (Optional)
    if (this.memberCount !== null) {
        ctx.fillStyle = this.memberCountColor;
        ctx.font = `medium 20px ${registeredFontName}-Medium`;
        ctx.textAlign = "end";
        ctx.fillText(`Member #${this.memberCount}`, cardWidth - padding, cardHeight - padding - 5);
    }

    // --- Encode and Return Buffer ---
    try {
      return await encodeToBuffer(canvas);
    } catch (err) {
      console.error("Failed to encode Welcome/Leave Alt card:", err);
      throw new Error("Could not generate the Welcome/Leave Alt card image buffer.");
    }
  }
};
